grpc/interceptors.go

```diff
@@ -24,6 +24,20 @@ const (
 	serverLatencyKey       = "server-latency"
 )
 
+// NoCancelInterceptor is a gRPC interceptor that creates a new context,
+// separate from the original context, that has the same deadline but does
+// not propagate cancellation. This is used by SA.
+func NoCancelInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+	cancel := func() {}
+	if deadline, ok := ctx.Deadline(); ok {
+		ctx, cancel = context.WithDeadline(context.Background(), deadline)
+	} else {
+		ctx = context.Background()
```

Comment:
Especially with honeycomb (or tracing in general) there's now a bunch of other info on this incoming context that we would rather not throw away. It's possible (not with gorp, but in general) to even propagate the trace context all the way to the mariadb server and have it emit spans of its own. Replacing the context entirely prevents us from doing that.

Even if we are willing to drop that info, we have to ensure that this interceptor runs after (or is wrapped by, however you prefer to think about it) the honeycomb interceptor, so that honeycomb can get a look at the incoming context before it gets replaced. I see that the implementation does do this, but it seems like it should be documented so this assumption doesn't get broken at a later date.

Reply:
As in, the MariaDB server has some sort of honeycomb or OpenTelemetry module? That's really cool!

My thinking is to use this to unblock our upgrade to that latest mysql driver, and also test our hypothesis that propagation of cancels is behind the performance regression. If that turns out the be the case, we can dig more into why it happens, and if there are any patches we can send upstream that would let us have both cancellation and our original performance.

If we do wind up having to keep this long term, is there any way to copy all the metadata from one context to another?

Reply:
Yeah, I think that this is a reasonable way to get the extra insight we want right now.

It definitely gives me pause: throwing away the context data means even the SA itself has no way to really know if a request got canceled, and it means we can't manually create sub-spans for any work the SA does (e.g. an SA gRPC call that results in three consecutive database calls; it might be helpful to have those wrapped in their own spans, and this prevents doing that).

So I think we should file a ticket to figure out a better way to do this (is there any way to create the new context immediately before firing off the request to the DB, instead of immediately upon receiving the request via gRPC?), but we should also proceed with this for now so we're not blocked on perfection.